### PR TITLE
WebP/Lossy: Implement macroblock coefficient decoding

### PR DESCRIPTION
**Pull request by nico**
This basically adds the line

    u8 token = TRY(
        tree_decode(decoder, COEFFICIENT_TREE,
        header.coefficient_probabilities[plane][band][tricky],
        last_decoded_value == DCT_0 ? 2 : 0));

and calls it once for the 16 coefficients of a discrete cosine transform
that covers a 4x4 pixel subblock.

And then it does this 24 or 25 times per macroblock, for the 4x4 Y
subblocks and the 2x2 each U and V subblocks (plus an optional Y2 block
for some macroblocks).

It then adds a whole bunch of machinery to be able to compute `plane`,
`band`, and in particular `tricky` (which depends on if the
corresponding left or above subblock has non-zero coefficients).

(It also does dequantization, and does an inverse Walsh-Hadamard
transform when needed, to end up with complete DCT coefficients
in all cases.)

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossyTables.h [hunk 1]
```diff
@@ -237,6 +237,25 @@ enum DCTToken {
...
+      -DCT_3, -DCT_4,       /* 3   = "111010", 4 = "111011" */
+     14, 16,
+      -dct_cat1, -dct_cat2, /* cat1 =  "111100",
+                               cat2 = "111101" */
+     18, 20,
+      -dct_cat3, -dct_cat4, /* cat3 = "1111100",
+                               cat4 = "1111101" */
+      -dct_cat5, -dct_cat6  /* cat4 = "1111110",
+                               cat4 = "1111111" */
+};
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossyTables.h:256]: I understand that clang-format will not give you two column styling,
but the inconsistent indenting still looks bad

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 2]
```diff
@@ -490,15 +494,430 @@ ErrorOr<Vector<MacroblockMetadata>> decode_VP8_macroblock_metadata(BooleanDecode
...
+}
+
+// Reading macroblock coefficients requires needing to know if the block to the left and above the current macroblock
+// has non-zero coefficients. This stores that state.
+struct CoefficientReadingContext {
+    // Store if each plane has nonzero coefficients in the block above and to the left of the current block.
+    Vector<bool> y2_above;
+    Vector<bool> y_above;
+    Vector<bool> u_above;
+    Vector<bool> v_above;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:659]: Can we make these `AK::Bitmaps`?

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossyTables.h (hunk 1 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossyTables.h:256]: I think the indentation is matching the structure of the tree?

Comment (Hendiadyoin1) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossyTables.h:256]: Oh, I see, that could be it

Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossyTables.h:256]: Right, it's in a clang-format off section to preserve the formatting from the spec. This is supposed to look like a tree.

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp (hunk 2 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:659]: There are `ceil_div(width * 9, 16)` of these, so 576 bytes for a 1024 wide image.

A 1024-wide image needs 4096 bytes to store the first scanline. Assuming it's just 10 pixel high, the 576 * 7/8 = 504 bytes savings are ~1% of the size needed for the image, and this is transient memory, only used during decoding.

So not sure this is worth worrying about.

**Comment by nico** [comment]
Red CI is unrelated.

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 3]
```diff
@@ -510,21 +929,20 @@ ErrorOr<NonnullRefPtr<Bitmap>> decode_webp_chunk_VP8_contents(VP8Header const& v
...
-#if 0
-    return Bitmap::create(bitmap_format, { vp8_header.width, vp8_header.height });
-#else
-    // FIXME: Implement webp lossy decoding.
-    (void)bitmap_format;
-    return Error::from_string_literal("WebPImageDecoderPlugin: decoding lossy webps not yet implemented");
-#endif
+    TRY(decode_VP8_image_data(*bitmap, header, vp8_header.second_partition, macroblock_width, macroblock_height, macroblock_metadata));
+
+    if (bitmap->physical_width() == (int)vp8_header.width && bitmap->physical_height() == (int)vp8_header.height)
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:943]: `bitmap->physical_size() == { ..., ... }`?

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 4]
```diff
@@ -510,21 +929,20 @@ ErrorOr<NonnullRefPtr<Bitmap>> decode_webp_chunk_VP8_contents(VP8Header const& v
...
+        return bitmap;
+    return bitmap->cropped({ 0, 0, (int)vp8_header.width, (int)vp8_header.height });
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:945]: Maybe extract width/height as variables and use C++-style casts

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 5]
```diff
@@ -490,15 +494,430 @@ ErrorOr<Vector<MacroblockMetadata>> decode_VP8_macroblock_metadata(BooleanDecode
...
+        return 2;
+    if (have_y2)
+        return 0;
+    return 3;
+}
+
+ErrorOr<i16> coefficient_value_for_token(BooleanDecoder& decoder, u8 token)
+{
+    // Implements the second half of https://datatracker.ietf.org/doc/html/rfc6386#section-13.2 "Coding of Individual Coefficient Values"
+    i16 v = (i16)token; // For DCT_0 to DCT4
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:560]: Please use a C++-style cast

**Comment by nico** [comment]
Thanks, all done!

**Pull request merged**